password_enhancements.module

```diff
@@ -0,0 +1,471 @@
+    $constraints_count = count($constraints);
+    $minimum_required_constraints = $policy->getMinimumRequiredConstraints();
+    $minimum_required_constraints = $constraints_count < $minimum_required_constraints ? $constraints_count : $minimum_required_constraints;
+
+    // Validate password against the defined constraints.
+    $error_message_collection = [
+      'optional' => [],
+      'required' => [],
+    ];
+    $valid_required = 0;
+    $valid_optional = 0;
+    $plugin_manager = Drupal::service('plugin.manager.password_constraint');
+    foreach ($constraints as $constraint) {
+      /** @var \Drupal\password_enhancements\Plugin\PasswordConstraintPluginInterface $plugin_instance */
+      $plugin_instance = $plugin_manager->createInstance($constraint->getType(), $constraint->getConfiguration());
+      $is_required = $constraint->isRequired();
+      $is_valid = $plugin_instance->validate($password);
+
+      if ($is_valid && $is_required) {
+        $valid_required++;
+      }
+      elseif ($is_valid) {
+        $valid_optional++;
+      }
+      else {
+        $error_message_collection[$is_required ? 'required' : 'optional'][] = $plugin_instance->getErrorMessage();
+      }
+    }
+
+    // Collect error messages.
+    $error_messages = [];
+    if (($valid_required + $valid_optional) < $minimum_required_constraints) {
+      $error_messages = array_merge($error_message_collection['required'], $error_message_collection['optional']);
+    }
+    elseif ($valid_required === 0 && $error_message_collection['required'] === [] && $valid_optional < $minimum_required_constraints) {
+      $error_messages = &$error_message_collection['optional'];
+    }
+    elseif ($error_message_collection['required'] !== []) {
+      $error_messages = &$error_message_collection['required'];
+    }
+
+    // Set form error and show the error message if there is any.
+    if ($error_messages !== []) {
+      $messages = [
+        '#theme' => 'password_enhancements_error_messages',
+        '#title' => t("The given password doesn't meet some requirements:"),
+        '#messages' => $error_messages,
+        '#attributes' => [
+          'class' => ['password-error-messages'],
+        ],
+        '#title_attributes' => [
+          'class' => ['title', 'password-error-message'],
+        ],
+      ];
+      $form_state->setError($form['account']['pass'], Drupal::service('renderer')->render($messages));
+    }
+  }
+}
+
+/**
+ * Submit callback for password_enhancements_form_user_register_form_alter().
+ *
+ * This callback runs only if the password field was set to always require and
+ * the verify_email set to true.
+ *
+ * @param array $form
+ *   The form.
+ * @param \Drupal\Core\Form\FormStateInterface $form_state
+ *   The current form state.
+ */
+function password_enhancements_form_user_register_form_alter_password_required_submit(array &$form, FormStateInterface $form_state): void {
```

Comment:
This extra submit callback is required because if the verify_email is set to
TRUE then by default Drupal generates a random password so we have to make
sure the password from the input is being set instead of the random generated
password.

I've upaded the doxygen with the same comment.